src/dd/dd.rs

```diff
@@ -0,0 +1,734 @@
+                                "cbs" => cf.cnv_block_size = m,
+                                "skip" => cf.skip = m,
+                                "seek" => cf.seek = m,
+                                "count" => cf.count = m,
+                                _ => err(format!("invalid argument: {}", key)),
+                            }
+                        }
+                    } else {
+                        err(format!("invalid argument: {}", arg));
+                    }
+                }
+            }
+        } else {
+            err(format!("invalid argument: {}", arg));
+        }
+    }
+
+    // Behaviour is unspecified. We choose to terminate.
+    if cf.count == 0 {
+        err("assign count non-zero value");
+    }
+
+    if cf.is_stdout && cf.seek != 0 {
+        err("cannot seek on standard output");
+    }
+
+    if cf.has_flag(Conf::BLOCK) && cf.has_flag(Conf::UNBLOCK) {
+        err("'block' and 'unblock' values are mutually exclusive");
+    }
+
+    if cf.has_flag(Conf::LCASE) && cf.has_flag(Conf::UCASE) {
+        err("'lcase' and 'ucase' values are mutually exclusive");
+    }
+
+    if !(cf.flags & Conf::BLOCK != 0 || cf.flags & Conf::UNBLOCK != 0 || cf.flags & Conf::LCASE != 0
+        || cf.flags & Conf::UCASE != 0 || cf.flags & Conf::SWAB != 0)
+        {
+            cf.short_to_short = true
+        }
+
+    // Behaviour is unspecified. We choose to terminate.
+    if cf.cnv_block_size == 0 && (cf.has_flag(Conf::BLOCK) || cf.has_flag(Conf::UNBLOCK)) {
+        err("assign cbs non-zero value");
+    }
+
+    cf
+}
+
+// , _sdone: chan::Sender<()>
+fn setup(cf: Arc<Mutex<Conf>>) -> (InpFileWrap, OutFileWrap) {
+    let mut cf = cf.lock().unwrap();
+
+    let inp = if cf.is_stdin {
+        Some(InpFileWrap {
+            f: InpFileType::Stdin(io::stdin()), // TODO: use .lock() if necessary
+            is_seekable: false,
+        })
+    } else {
+        let f = File::open(cf.inp_name.as_ref() as &str);
+        if f.is_err() {
+            err("cannot open input file".to_string());
+            None
+        } else {
+            Some(InpFileWrap {
+                f: InpFileType::Reg(f.unwrap()),
+                is_seekable: true, // TODO: handle case when device isn't seekable
+            })
+        }
+    };
+    let mut inp = inp.unwrap();
+
+    let out = if cf.is_stdout {
+        Some(OutFileWrap {
+            f: OutFileType::Stdout(io::stdout()), // TODO: use .lock() if necessary
+            is_seekable: false,
+        })
+    } else {
+        let f = File::create(cf.out_name.as_ref() as &str);
+        if f.is_err() {
+            err("cannot open output file".to_string());
+            None
+        } else {
+            Some(OutFileWrap {
+                f: OutFileType::Reg(f.unwrap()),
+                is_seekable: true, // TODO: handle case when device isn't seekable
+            })
+        }
+    };
+    let mut out = out.unwrap();
+
+    match inp.f {
```

Comment:
Can you move this and the `match out.f` into separate functions?

Reply:
Moved.